### PR TITLE
fix(gitutil): carry a rebase through every conflicting commit

### PR DESCRIPTION
**Pull request by rsnodgrass**
## Summary

Follow-up to #721. That PR made `sessions/` conflicts resolvable; this one makes the resolver actually finish a **multi-commit** rebase. Verified by healing a copy of the real production ledger: **350 ahead / 1067 behind → 344 ahead / 0 behind in 28 seconds.**

## What was still broken

A rebase replays commits **one at a time**, so `git rebase --continue` exits **non-zero every time** it commits the current step and halts on the next conflicting commit. That is progress, not failure — but the resolver treated any non-zero exit as fatal:

```
resolve step 1 → rebase --continue → exit 1 → return error → caller aborts
                                   → pre-rebase state restored → wedged again
```

Captured directly from the real ledger:

```
exit code: 1
[detached HEAD 0adea3c09] ox doctor: auto-commit ledger changes
 587 files changed, 159927 insertions(+), 887 deletions(-)
--- after: rebase still in progress? YES, unmerged=1
```

Non-zero exit, rebase still running, one fresh conflict waiting. The resolver bailed and the caller aborted — on every single attempt, forever.

**This cannot reproduce with a single conflicting commit**, which is exactly why the existing fixtures (and my own first attempt at a regression test) passed while a 344-commit ledger stayed stuck for 13 days.

## The fix

`ResolveRebaseAcceptTheirs` now loops over rebase steps instead of handling one. Exit code is no longer trusted as the signal — **repo state is**:

- rebase no longer in progress → done
- `--continue` succeeded, still mid-rebase → advance to the next step
- `--continue` errored, still mid-rebase, **fresh unmerged files present** → that is the next conflict; keep going
- `--continue` errored with nothing to resolve → genuine failure, surface it (git wants something we cannot act on, and looping would spin)

Bounded by `maxResolvePasses` (5000) so a pass that resolves nothing can never spin forever. Error output goes through `SanitizeOutput` — git errors embed the PAT and local paths.

## Test Plan

- **`TestWedge_SequentiallyConflictingCommits`** — 10 local commits each conflicting on their *own* session file, so the replay halts at every step. One resolver call must carry it through and leave no rebase in progress. **Verified red-first**: restoring the original `if contErr != nil { return error }` makes it fail; the fix makes it pass.
- **`TestWedge_ReplayedCommitBecomesEmpty`** — the adjacent shape, where a replayed commit's changes are already upstream and git halts with no unmerged files.
- Both join the existing wedge harness and assert the same three-part contract: detected → cleared → **the ledger actually pulls and pushes afterward**.
- **Real-ledger proof**: a copy of the production ledger (350/1067, 47 merge-tree conflicts) reconciles to 344 ahead / **0 behind** with the rebase fully complete. The original was never mutated.
- `internal/gitutil`, `internal/lfs`, `internal/ledger`, `internal/daemon`, `cmd/ox` all green; `make lint` 0 issues.

## Why my first regression test didn't catch it

Worth recording, because it is the same trap that let the bug ship. My first fixture used 10 commits all touching the *same* file — git collapsed them into a single conflict and one `--continue` finished the job, so the test passed with the bug still in place. Only when each commit conflicts on a **distinct** file does the replay halt repeatedly and reproduce the failure. A regression test that passes both with and without the fix guards nothing, which is why every case here is verified red-first.

---
<details>
<summary>Checklist (expand if needed)</summary>

- [x] Tests added, verified red-first
- [ ] CHANGELOG — N/A, internal repair path
- [x] Breaking change — none
- [x] `/security-review` — N/A for this diff (no auth, MCP, raw_writer, prepush_scan, upgrade, or go.sum changes)

</details>

Co-Authored-By: [SageOx](https://github.com/SageOx)

## Summary by CodeRabbit

- **Bug Fixes**
  - Ledger doctor now properly identifies “unborn” ledgers (no commits) instead of treating them as detached, with clearer critical outcomes for empty remotes, unreachable remotes, and repair behavior.
  - Rebase “accept theirs” repair now advances through multi-step conflict resolution more reliably, including edge cases like empty halts.
- **Chores**
  - CI runs now cancel in-progress non-main builds; local test preflight runs lint/tests in parallel.
  - Lint configuration updated to narrow a specific test-file staticcheck exclusion.
- **Tests**
  - Added and parallelized regression coverage for unborn-ledger and rebase/wedge scenarios.

### CODE REVIEW

File: internal/gitutil/rebase_resolve.go [hunk 1]
```diff
@@ -37,10 +70,10 @@ func ResolveRebaseAcceptTheirs(ctx context.Context, repoPath string, safePrefixe
 	// list all unmerged files from the index (handles content, rename, and delete conflicts)
 	entries, err := listUnmergedEntries(ctx, repoPath)
 	if err != nil {
-		return fmt.Errorf("list unmerged entries: %w", err)
+		return false, fmt.Errorf("list unmerged entries: %w", err)
 	}
 	if len(entries) == 0 {
-		return fmt.Errorf("no conflicted files found")
+		return false, fmt.Errorf("no conflicted files found")
 	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/gitutil/rebase_resolve.go:88]: Fixed in af559380. You are right about the code path, and the way you are wrong about the test is the interesting part.

**The bug is real.** `resolveOneRebaseStep` returned `no conflicted files found` whenever the index had no conflicts — including while a rebase was still ACTIVE. During an active rebase that is not an error at all: git halts that way when a replayed commit's changes are already upstream. Returning an error made the caller abort, restore the pre-rebase state, and re-wedge the ledger. Same failure shape as the multi-commit bug this PR fixes, one layer down.

**The test claim does not hold, and that is worth recording.** `TestWedge_ReplayedCommitBecomesEmpty` passes on this branch, before and after. I checked why rather than assuming either of us was right: git 2.55 defaults to `--empty=drop` for non-interactive rebase, so an emptied commit is silently dropped and the rebase never halts. I also tried forcing it with `-c rebase.empty=stop` and it still auto-dropped. So the fixture cannot reach your path on modern git — the code was broken while every end-to-end test stayed green. That is exactly the trap that let the original 13-day wedge ship, so I did not want to "fix" it on a test that could not fail.

**What changed:**
- In-rebase with zero unmerged entries now routes to `advanceNonConflictRebaseStep` instead of erroring.
- That helper tries `--continue` first and escalates to `--skip` ONLY when git itself names the remedy (`--skip` / `nothing to commit` / `patch is empty` / `previous cherry-pick is now empty`). It never skips blindly — `--skip` DROPS the current commit, so guessing there would silently discard a real session, which is worse than the wedge.
- Anything else that halts with nothing to resolve still returns an error, sanitized through `SanitizeOutput` (git errors embed the PAT and local paths).
- The main `--continue` block now shares the same helper, so both paths agree.

**Tests**, both verified red-first:
- `TestWedge_RebaseHaltsWithNothingToResolve` — constructs the state directly (resolve and stage by hand mid-rebase) rather than relying on git to produce it, so it works regardless of git's `--empty` default. Restoring the old unconditional error makes it fail with "a rebase halted with nothing to resolve must be advanced, not treated as an error".
- `TestResolveRebase_NoConflictsOutsideRebaseStillErrors` — pins the original contract, so the fix cannot silently turn a genuine caller error into a no-op.

`make lint` 0 issues; `internal/gitutil` green under `-race`; `internal/daemon`, `cmd/ox`, `internal/lfs` all green.

File: cmd/ox/doctor_ledger_git.go [hunk 2]
```diff
@@ -312,6 +326,66 @@ func checkLedgerBranchStatus(fix bool) checkResult {
...
+			return r
+		}
+		if out, err := exec.Command("git", "-C", ledgerPath, "fetch", "origin").CombinedOutput(); err != nil {
+			return FailedCheck(name, "fetch failed", gitutil.SanitizeOutput(strings.TrimSpace(string(out))))
+		}
+		if out, err := exec.Command("git", "-C", ledgerPath, "checkout", branch).CombinedOutput(); err != nil {
+			return FailedCheck(name, "checkout failed", gitutil.SanitizeOutput(strings.TrimSpace(string(out))))
+		}
+		return PassedCheck(name, fmt.Sprintf("restored branch %q from remote", branch))
+	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/doctor_ledger_git.go]: Fixed in 0bf88faa. This one is not hypothetical — it nearly bit me while repairing real ledgers earlier today.

One of the ledgers had an **expired PAT embedded in its origin URL**, so `ls-remote` failed with `HTTP Basic: Access denied`. Under the old code that is indistinguishable from "zero heads", so the check would have told me the ledger was never provisioned and handed me a `git commit -m 'seed ledger' && git push -u origin main` recipe. That ledger had real remote history. Following that guidance would have authored a divergent root commit on a shared team ledger.

There are now **three** outcomes instead of two:

| `ls-remote` | verdict | guidance |
|---|---|---|
| succeeds, has heads | local clone lost its branch | auto-repair: fetch + checkout |
| succeeds, zero heads | verifiably never provisioned | seeding is correct, and only here |
| **fails** | **cannot tell** | critical, and explicitly says *do not seed until the remote is reachable*; points at `ox doctor` / `ox login` |

Switched from `.Output()` to `.CombinedOutput()` so the failure reason is captured, and it goes through `gitutil.SanitizeOutput` before display — git auth errors embed the PAT and local paths, which is what made this failure mode invisible in the first place.

`TestUnbornLedger_UnreachableRemoteNeverSuggestsSeeding` covers both halves: an unreachable remote must stay critical AND must not contain "seed ledger" in its guidance; a reachable-but-empty remote must still offer it.

File: cmd/ox/doctor_ledger_git.go [hunk 3]
```diff
@@ -312,6 +326,66 @@ func checkLedgerBranchStatus(fix bool) checkResult {
...
+			r.slug = CheckSlugLedgerBranchStatus
+			r.fixLevel = FixLevelAuto
+			return r
+		}
+		if out, err := exec.Command("git", "-C", ledgerPath, "fetch", "origin").CombinedOutput(); err != nil {
+			return FailedCheck(name, "fetch failed", gitutil.SanitizeOutput(strings.TrimSpace(string(out))))
+		}
+		if out, err := exec.Command("git", "-C", ledgerPath, "checkout", branch).CombinedOutput(); err != nil {
+			return FailedCheck(name, "checkout failed", gitutil.SanitizeOutput(strings.TrimSpace(string(out))))
+		}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/doctor_ledger_git.go]: Fixed in 0bf88faa. Correct on both counts — the demotion and the lost slug.

`FailedCheck` leaves `priority` empty, and `categorizeCheck` sends anything non-critical to the `attention` bucket, so a FAILED repair of a never-synced ledger ended up quieter than the critical detection that triggered it. That is exactly backwards: a repair that could not complete is strictly worse news than the finding, because now the user also knows automation cannot fix it.

Every non-passing return from `unbornLedgerFailure` now goes through one small `critical()` helper that sets `priority = "critical"` and `slug = CheckSlugLedgerBranchStatus`. Using a helper rather than patching the three call sites individually so a future early-return cannot silently skip it — that is how the original gap appeared.

`TestUnbornLedger_FailedRepairStaysCritical` asserts a failed repair keeps both the critical priority and the slug.

File: internal/gitutil/rebase_resolve.go [hunk 4]
```diff
@@ -135,41 +179,101 @@ func ResolveRebaseAcceptTheirs(ctx context.Context, repoPath string, safePrefixe
...
+		return false, nil // advanced to the next step
+	}
+
+	// git refuses --continue on an empty step and names the remedy. Only then
+	// is dropping the commit the correct (and git-sanctioned) action.
+	lower := strings.ToLower(contOut)
+	emptyStep := strings.Contains(lower, "--skip") ||
+		strings.Contains(lower, "nothing to commit") ||
+		strings.Contains(lower, "patch is empty") ||
+		strings.Contains(lower, "previous cherry-pick is now empty")
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/gitutil/rebase_resolve.go]: Fixed in 7d0b3105. Correct, and the failure mode you describe is the silent kind — a mistranslated hint would make us report "nothing to resolve", the caller would abort, and the ledger would re-wedge with no indication why.

Rather than translate the strings, the check is now **structural**. A replayed commit is empty exactly when nothing is staged and nothing is conflicted — git has applied the changes and found no delta:

```go
func rebaseStepIsEmpty(ctx, repoPath) bool {
    if unmerged, err := listUnmergedEntries(ctx, repoPath); err != nil || len(unmerged) > 0 {
        return false
    }
    _, err := RunGit(ctx, repoPath, "diff", "--cached", "--quiet")
    return err == nil // exit 0 == index matches HEAD == nothing to commit
}
```

That is language-independent and does not depend on git's wording staying put.

Two supporting changes:
- `runRebaseStep` now pins `LC_ALL=C` / `LANG=C`, so any diagnostic we *do* read is stable regardless of the user's locale.
- The old substring list survives only as `mentionsEmptyStep`, and it is consulted **only when the structural check says "not empty"** — so a translation gap can never be the sole reason we drop a commit. That ordering matters: `--skip` DISCARDS the current commit, so the dangerous direction is a false positive, and the structural check is the one that gates it.

File: internal/gitutil/rebase_resolve.go [hunk 5]
```diff
@@ -135,41 +179,130 @@ func ResolveRebaseAcceptTheirs(ctx context.Context, repoPath string, safePrefixe
...
+	// "nothing to resolve", the caller would abort, and the ledger would re-wedge.
+	// runRebaseStep pins LC_ALL=C so the text is stable, but the structural check
+	// is what we actually trust; the text is only a fallback for odd git builds.
+	//
+	// A replayed commit is empty exactly when nothing is staged and nothing is
+	// conflicted: git has already applied the changes and found no delta.
+	if !rebaseStepIsEmpty(ctx, repoPath) && !mentionsEmptyStep(contOut) {
+		return false, fmt.Errorf("rebase halted with nothing to resolve: %s",
+			SanitizeOutput(strings.TrimSpace(contOut)))
+	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply internal/gitutil/rebase_resolve.go]: Fixed — took your suggestion exactly. The text fallback is gone.

You are right that it was strictly unsafe, and my "belt and braces" reasoning for adding it was backwards. A pre-commit hook that rejects `--continue` while real changes are staged can trivially emit `--skip` or `nothing to commit` in its OWN output, and that would have overridden the structural check and dropped a genuine commit.

The asymmetry is what settles it: `--skip` DISCARDS work, so the only tolerable error direction is refusing to skip. A false negative costs one loud error; a false positive costs a session permanently. `rebaseStepIsEmpty` is now the sole signal, and `mentionsEmptyStep` is deleted rather than left around to be re-wired by someone later. `LC_ALL=C` stays on `runRebaseStep` so anything we surface in an error message is stable, but nothing reads that text to make a decision anymore.

File: cmd/ox/doctor_ledger_git.go [hunk 6]
```diff
@@ -312,6 +326,91 @@ func checkLedgerBranchStatus(fix bool) checkResult {
...
+		if !fix {
+			r := critical(CriticalCheck(name,
+				fmt.Sprintf("branch %q has no commits (remote does)", branch),
+				fmt.Sprintf("The local clone lost its branch — %d uncommitted file(s) have never synced.\n       "+
+					"Run `ox doctor --fix` to restore it from the remote.", untracked)))
+			r.fixLevel = FixLevelAuto
+			return r
+		}
+		if out, err := exec.Command("git", "-C", ledgerPath, "fetch", "origin").CombinedOutput(); err != nil {
+			return critical(FailedCheck(name, "fetch failed", gitutil.SanitizeOutput(strings.TrimSpace(string(out)))))
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/doctor_ledger_git.go]: Fixed. Restructured to checkout-first, fetch-only-if-needed.

A clone that merely lost its branch ref almost always still HAS the objects, so `git checkout <branch>` succeeds with no network at all. That is now the first attempt, and in the common case `ox doctor --fix` never touches the daemon's fetch lane.

The fetch remains only as the narrow fallback when the objects are genuinely absent, and it is now scoped to `fetch origin <branch>` rather than a bare `fetch origin`, so it cannot contend with the daemon over unrelated refs. A ledger that has never synced at all is worth one targeted fetch to recover.

For context on the contract: `fixLedgerBranchDiverged` in this same file already fetches under `--fix` (it re-checks ahead/behind before the heavier push path), so repair paths fetching is pre-existing precedent rather than something this PR introduces — but your point stands that the default should be to avoid it, and now it does.

File: cmd/ox/doctor_ledger_git_test.go [hunk 7]
```diff
@@ -299,3 +299,173 @@ func TestStripURLCredentials(t *testing.T) {
...
+	// remote has commits, so the repair path runs — then break it by pointing
+	// origin somewhere unfetchable AFTER the ls-remote succeeds is not possible
+	// in-process, so assert the shape of a checkout failure on a bogus branch.
+	res := unbornLedgerFailure(repo, "no-such-branch", true)
+
+	if !res.passed {
+		assert.Equal(t, "critical", res.priority,
+			"a failed repair must stay critical, not be demoted to attention")
+		assert.Equal(t, CheckSlugLedgerBranchStatus, res.slug)
+	}
```
[the remark this answers is not in the document]
Comment (rsnodgrass) [reply cmd/ox/doctor_ledger_git_test.go]: Fixed. Both points were right, and the first one is explicitly called out in this repo's own `.claude/rules/testing.md` under anti-patterns ("Wrapping assertions in conditionals — passes vacuously"). I wrote exactly that. 

- The `if !res.passed` guard is gone. It now asserts `require.False(t, res.passed, ...)` unconditionally, so the test fails loudly if the repair is ever skipped and returns success — which was precisely the hole you identified.
- Fixture setup no longer discards git errors: the `run` helper is `require.NoError(t, err, "fixture git %v failed: %s", args, out)`, so a broken fixture fails immediately instead of leaving the test asserting against a repo in an unintended state.
- Also made the fixture unambiguous — the branch under repair (`no-such-branch`) now exists nowhere, so both the local checkout and the fetch must fail and the failure path is guaranteed to be exercised.
